internal/services/ingest.go

```diff
@@ -155,6 +169,212 @@ func (m *ingestService) Run(ctx context.Context, startLedger uint32, endLedger u
+	if inSync {
+		return GetLedgersResponse{}, ErrAlreadyInSync
+	}
+
+	getLedgersResponse, err := m.rpcService.GetLedgers(ledgerSeqRange.StartLedger, ledgerSeqRange.Limit)
+	if err != nil {
+		return GetLedgersResponse{}, fmt.Errorf("getting ledgers: %w", err)
+	}
+
+	return getLedgersResponse, nil
+}
+
+type LedgerSeqRange struct {
+	StartLedger uint32
+	Limit       uint32
+}
+
+// getLedgerSeqRange returns a ledger sequence range to ingest. It takes into account:
+// - the ledgers available in the RPC,
+// - the latest ledger synced by the ingestion service,
+// - the max ledger window to ingest.
+//
+// The returned ledger sequence range is inclusive of the start and end ledgers.
+func getLedgerSeqRange(rpcOldestLedger, rpcNewestLedger, latestLedgerSynced uint32) (ledgerRange LedgerSeqRange, inSync bool) {
+	if latestLedgerSynced >= rpcNewestLedger {
+		return LedgerSeqRange{}, true
+	}
+	ledgerRange.StartLedger = max(latestLedgerSynced+1, rpcOldestLedger)
+	ledgerRange.Limit = getLedgersLimit
+
+	return ledgerRange, false
+}
+
+func (m *ingestService) processLedgerResponse(ctx context.Context, getLedgersResponse GetLedgersResponse) error {
+	// Create a worker pool with
+	const poolSize = 16
+	pool := pond.New(poolSize, len(getLedgersResponse.Ledgers), pond.Context(ctx))
+
+	// Create a slice of errors
+	mu := sync.Mutex{}
+	var allTxHashes []string
+	var errs []error
+
+	// Submit tasks to the pool
+	for _, ledger := range getLedgersResponse.Ledgers {
+		ledger := ledger // Create a new variable to avoid closure issues
+		pool.Submit(func() {
+			txHashes, err := m.processLedger(ctx, ledger)
+			mu.Lock()
+			if err != nil {
+				errs = append(errs, fmt.Errorf("processing ledger %d: %w", ledger.Sequence, err))
+			} else {
+				allTxHashes = append(allTxHashes, txHashes...)
+			}
+			mu.Unlock()
+		})
+	}
+
+	// Wait for all tasks to complete
+	pool.StopAndWait()
+
+	if len(errs) > 0 {
+		return fmt.Errorf("processing ledgers: %w", errors.Join(errs...))
+	}
+
+	// TODO: ingest processed data
+	// TODO: unlock channel accounts
+
+	log.Ctx(ctx).Infof("🚧 Done processing & ingesting %d ledgers", len(getLedgersResponse.Ledgers))
+
+	return nil
+}
+
+func (m *ingestService) processLedger(ctx context.Context, ledgerInfo protocol.LedgerInfo) ([]string, error) {
+	var xdrLedgerCloseMeta xdr.LedgerCloseMeta
+	if err := xdr.SafeUnmarshalBase64(ledgerInfo.LedgerMetadata, &xdrLedgerCloseMeta); err != nil {
+		return nil, fmt.Errorf("unmarshalling ledger close meta: %w", err)
+	}
+
+	transactions, err := m.getLedgerTransactions(ctx, xdrLedgerCloseMeta)
+	if err != nil {
+		return nil, fmt.Errorf("getting ledger transactions: %w", err)
+	}
+
+	// TODO: process transactions
+	log.Ctx(ctx).Debugf("🚧 Got %d transactions for ledger %d", len(transactions), xdrLedgerCloseMeta.LedgerSequence())
+	txHashes := make([]string, 0, len(transactions))
+	for _, tx := range transactions {
+		txHashes = append(txHashes, tx.Hash.HexString())
```

Comment:
Since this is still in TODO, I am assuming this will return a list of `types.Transaction` structs instead of tx hash strings? Because what would be the use of the tx hashes?